### ISSUE TITLE
real cases tested about your segmentation function

### ISSUE THREAD

**Issue by shukami1** [facebookresearch/dinov2#565]
Hello,

Thanks for your amazing work and sharing the repo with community.

I tested some real cases produced from our poor camera about your segmentation function. Results shown in the link below. Is the proper results?  If not, would you like give me some guide about the right answer? If yes, can you give me some suggestion?  THX
My codes written like this:
`def load_dino_model():
    """
    Load the DINOv2 model and segmentation head once.
    """
    global dino_model, dino_decoder, dino_transform
    
    import torch
    import torchvision.transforms as T
    import sys
    import os
    
    os.environ["XFORMERS_DISABLED"] = "1"
    
    # Add the dinov2 path to sys.path to import modules
    dinov2_path = '/home/xxxx/.cache/torch/hub/facebookresearch_dinov2_main'
    if dinov2_path not in sys.path:
        sys.path.insert(0, dinov2_path)
    
    # Import required modules from dinov2
    from dinov2.eval.segmentation.models.decode_heads.linear_head import BNHead
    
    # Load DINOv2 backbone model if not already loaded
    if dino_model is None:
        dino_model = torch.hub.load('/home/xxxx/.cache/torch/hub/facebookresearch_dinov2_main', 'dinov2_vitl14', source='local')
        dino_model.to(DEVICE).eval()
        print("DINOv2 model loaded")

    # Create ADE20K dataset linear segmentation head if not already created
    # ADE20K has 150 semantic classes 
    if dino_decoder is None:
        dino_decoder = BNHead(
            in_channels=1024,   # For vit-l/14
            channels=1024,      # Should match in_channels for BNHead
            num_classes=150,   # ADE20K has 150 classes (not 151)
            in_index=0         # Index of the input feature
        )
        
        # Load segmentation head weights
        checkpoint_path = 'checkpoints/dinov2_vitl14_ade20k_linear_head.pth'
        if os.path.exists(checkpoint_path):
            checkpoint = torch.load(checkpoint_path, map_location='cpu')
            
            # Extract state_dict from checkpoint
            if 'state_dict' in checkpoint:
                state_dict = checkpoint['state_dict']
            else:
                state_dict = checkpoint
            
            # Process weight dictionary
            new_state_dict = {}
            for k, v in state_dict.items():
                # Remove "decode_head." prefix
                if k.startswith('decode_head.'):
                    new_key = k[12:]  # Remove 'decode_head.' prefix
                    new_state_dict[new_key] = v
                else:
                    new_state_dict[k] = v
            
            # Load weights into segmentation head
            dino_decoder.load_state_dict(new_state_dict, strict=True)
            print(f"Loaded segmentation head from {checkpoint_path}")
        else:
            print(f"Segmentation head checkpoint not found at {checkpoint_path}")
            
        dino_decoder.to(DEVICE).eval()
    
    # Create transform if not already created
    if dino_transform is None:
        dino_transform = T.Compose([
            T.ToPILImage(),
            T.Resize((518, 518)),
            T.ToTensor(),
            T.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
        ])

def segment_with_dino(image):
    global dino_model, dino_decoder, dino_transform
    
    # Check if models are loaded
    if dino_model is None or dino_decoder is None or dino_transform is None:
        raise RuntimeError("DINO models not loaded. Call load_dino_model() first.")
    
    # Apply Gaussian blur to reduce compression artifacts and blocky structures
    # Using a small kernel size (5x5) with sigma=0.5 to minimally smooth the image
    # blurred_image = cv2.GaussianBlur(image, (5, 5), 3)
    blurred_image = image
    
    # Transform input
    img_tensor = dino_transform(blurred_image).unsqueeze(0).to(DEVICE)

    # Forward pass
    with torch.no_grad():
        # Get features
        features = dino_model.forward_features(img_tensor)
        # Extract patch features
        if isinstance(features, dict) and 'x_norm_patchtokens' in features:
            patch_tokens = features['x_norm_patchtokens']
        elif isinstance(features, dict) and 'x_norm' in features:
            patch_tokens = features['x_norm']
        else:
            patch_tokens = features

        # Reshape features to fit segmentation head
        h, w = img_tensor.shape[2] // 14, img_tensor.shape[3] // 14
        patch_tokens = patch_tokens.reshape(1, h, w, -1).permute(0, 3, 1, 2)
        
        # Apply segmentation head
        segmentation = dino_decoder([patch_tokens])
        
        # Get prediction result
        seg_pred = segmentation.argmax(dim=1).squeeze().cpu().numpy()
        
        # Resize to match original image
        seg_resized = cv2.resize(seg_pred.astype(np.uint8), (image.shape[1], image.shape[0]), 
                                interpolation=cv2.INTER_NEAREST)
        
        # Create colorized segmentation map
        seg_color = cv2.applyColorMap((seg_resized * 2).astype(np.uint8), cv2.COLORMAP_JET)

    return seg_resized, seg_color`

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/6fa22dea-7f5f-4825-8f68-cceee3b837a2" />
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/363dbbd6-e18f-4b32-85b6-0f918cd33b60" />
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/e6166591-a7c7-4a5c-b92a-61a3441b606f" />
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/253d693c-c0ba-4d22-8d6f-a44dd3b6ef54" />
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/34ca6578-83c8-497c-af19-54411306b36b" />
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/bf1d5261-2a86-4060-b9c7-aff4ed28d2d4" />
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/8715df09-4c13-4602-a9f8-752ad24e679f" />
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/4191042c-0fae-42d5-8d17-3a0604dc143b" />
